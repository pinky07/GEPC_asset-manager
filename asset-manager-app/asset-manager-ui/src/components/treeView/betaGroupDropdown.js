import {
  Dropdown,
  DropdownItem,
  DropdownMenu,
  DropdownToggle,
} from 'reactstrap';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import React from 'react';

import { getBetaGroups, selectBetaGroup } from '../../actions';

export class BetaGroupDropdown extends React.Component {
  static defaultProps = {
    selectedBetaGroup: {
      id: -1,
      name: 'Beta Groups',
    },
  };

  constructor(props) {
    super(props);
    this.state = {
      dropdownOpen: false,
    };
  }

  componentDidMount() {
    this.props.getBetaGroups();
  }

  dropdownToggle = () => {
    this.setState({
      dropdownOpen: !this.state.dropdownOpen,
    });
  };

  changeBetaGroup = event => {
    const group = {
      id: parseInt(event.currentTarget.value, 10),
      name: event.currentTarget.textContent,
    };
    this.props.selectBetaGroup(group);
  };

  renderDropdownItems = () => {
    if (this.props.betaGroups) {
      return (
        <DropdownMenu>
          {this.props.betaGroups.map(betaGroup => (
            <DropdownItem
              key={betaGroup.id}
              value={betaGroup.id}
              onClick={this.changeBetaGroup}>
              {betaGroup.name}
            </DropdownItem>
          ))}
        </DropdownMenu>
      );
    }
    return null;
  };

  render() {
    return (
      <Dropdown isOpen={this.state.dropdownOpen} toggle={this.dropdownToggle}>
        <DropdownToggle caret>
          {this.props.selectedBetaGroup.name}
        </DropdownToggle>
        {this.renderDropdownItems()}
      </Dropdown>
    );
  }
}

BetaGroupDropdown.propTypes = {
  selectedNode: PropTypes.object,
  selectedBetaGroup: PropTypes.object,
  betaGroups: PropTypes.array.isRequired,
  selectBetaGroup: PropTypes.func.isRequired,
  getBetaGroups: PropTypes.func.isRequired,
};

const mapStateToProps = state => {
  return {
    selectedBetaGroup: state.allocationTree.selectedBetaGroup,
    betaGroups: state.allocationTree.betaGroups,
  };
};

export default connect(mapStateToProps, {
  getBetaGroups,
  selectBetaGroup,
})(BetaGroupDropdown);
