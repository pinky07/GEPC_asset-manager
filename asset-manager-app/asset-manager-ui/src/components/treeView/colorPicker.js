import PropTypes from 'prop-types';
import React from 'react';
import reactCSS from 'reactcss';
import { SketchPicker } from 'react-color';

const DEFAULT_COLOR_PICKER = '#002060';
const PRESET_COLORS = [
  '#a0ac1b',
  '#4d4e54',
  '#eeece1',
  '#d0d68f',
  '#a0bdc0',
  '#993300',
  '#cc9900',
  '#557d81',
  '#da7d00',
  '#fce7b2',
  '#666635',
  '#feae0e',
  '#640000',
  '#002060',
];

export default class ColorPicker extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      displayColorPicker: false,
      selectedColor: DEFAULT_COLOR_PICKER,
      presetColors: PRESET_COLORS,
    };
  }

  onClickColorPicker = () => {
    if (!this.props.disabled) {
      this.setState({ displayColorPicker: !this.state.displayColorPicker });
    }
  };

  onClose = () => {
    this.props.changeColor(this.state.selectedColor);
    this.setState({ displayColorPicker: false });
  };

  onChangeColor = color => {
    this.setState({ selectedColor: color.hex });
  };

  renderPicker() {
    if (this.state.displayColorPicker) {
      return (
        <div className="popover">
          <div className="cover" onClick={this.onClose} />
          <SketchPicker
            color={this.state.selectedColor}
            onChange={this.onChangeColor}
            presetColors={this.state.presetColors}
          />
        </div>
      );
    }
    return null;
  }

  render() {
    const styles = reactCSS({
      default: {
        color: {
          background: this.state.selectedColor,
        },
      },
    });

    return (
      <div className="colorPicker">
        <div onClick={this.onClickColorPicker}>
          <div className="color" style={styles.color} />
        </div>
        {this.renderPicker()}
      </div>
    );
  }
}

ColorPicker.propTypes = {
  disabled: PropTypes.bool,
  changeColor: PropTypes.func,
};
