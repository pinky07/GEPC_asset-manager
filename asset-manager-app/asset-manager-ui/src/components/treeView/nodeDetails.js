import {
  Card,
  CardBlock,
  CardTitle,
  Input,
  InputGroup,
  InputGroupAddon,
} from 'reactstrap';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import React from 'react';

import TreeService from '../../services/treeService';
import { updateDetailsNode } from '../../actions';
import ColorPicker from './colorPicker';

export class NodeDetails extends React.Component {
  constructor(props) {
    super(props);
    this.node = {
      policy_value: '',
      actual_mv: '',
      accountgroupshortname: '',
      accountgroupname: '',
    };
    this.state = {
      node: this.node,
    };
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.selectedNode) {
      const { node } = nextProps.selectedNode;
      this.setState({
        node,
      });
    } else {
      this.setState({
        node: this.node,
      });
    }
  }

  onChangeInput = (event, nodeProp) => {
    let node = { ...this.state.node };
    node[nodeProp] = event.target.value;
    if (nodeProp === 'accountgroupname') {
      node.title = node.accountgroupname;
    }
    this.setState({
      node,
    });
  };

  onChangePolicy(event) {
    this.setState({
      node: {
        ...this.state.node,
        policy_value: event.target.value,
      },
    });
  }

  onChangeActual(event) {
    this.setState({
      node: {
        ...this.state.node,
        actual_mv: event.target.value,
      },
    });
  }

  isRootNode = () => TreeService().isRootNode(this.props.selectedNode);

  renderPolicyValue = () =>
    this.state.node.policy_value ? `${this.state.node.policy_value}%` : '';

  onBlurTest = () => this.props.updateDetailsNode(this.state.node);

  changeColorAssignment = color => {
    this.setState(
      {
        node: {
          ...this.state.node,
          colorAssignment: color,
        },
      },
      () => this.props.updateDetailsNode(this.state.node)
    );
  };

  render() {
    return (
      <div className="nodeDetails">
        <Card>
          <CardBlock>
            <CardTitle>Node Details</CardTitle>
            <InputGroup>
              <InputGroupAddon>Node Name</InputGroupAddon>
              <Input
                value={this.state.node.accountgroupname}
                onChange={event =>
                  this.onChangeInput(event, 'accountgroupname')}
                onBlur={this.onBlurTest}
              />
            </InputGroup>
            <br />

            <InputGroup>
              <InputGroupAddon>Short Code</InputGroupAddon>
              <Input
                value={this.state.node.accountgroupshortname}
                disabled={this.isRootNode()}
              />
            </InputGroup>
            <br />

            <InputGroup>
              <InputGroupAddon>CRM Invest Prod</InputGroupAddon>
              <Input disabled={this.isRootNode()} />
            </InputGroup>
            <br />

            <InputGroup>
              <InputGroupAddon>Policy Weight</InputGroupAddon>
              <Input
                value={this.renderPolicyValue()}
                onChange={this.onChangePolicy}
                disabled={this.isRootNode()}
              />
            </InputGroup>
            <br />

            <InputGroup>
              <InputGroupAddon>Actual MV</InputGroupAddon>
              <Input
                value={this.state.node.actual_mv || ''}
                onChange={this.onChangeActual}
                disabled={this.isRootNode()}
              />
            </InputGroup>
            <br />

            <InputGroup>
              <InputGroupAddon>Assumption</InputGroupAddon>
              <Input disabled={this.isRootNode()} />
            </InputGroup>
            <br />

            <InputGroup>
              <InputGroupAddon>Return Series</InputGroupAddon>
              <Input disabled={this.isRootNode()} />
            </InputGroup>
            <br />

            <InputGroup>
              <InputGroupAddon>Color Assignment</InputGroupAddon>
              <ColorPicker
                changeColor={this.changeColorAssignment}
                disabled={this.isRootNode()}
              />
            </InputGroup>
            <br />

            <InputGroup>
              <InputGroupAddon>Liquidity</InputGroupAddon>
              <Input disabled={this.isRootNode()} />
            </InputGroup>
            <br />

            <InputGroup>
              <InputGroupAddon>Active/Pasive</InputGroupAddon>
              <Input disabled={this.isRootNode()} />
            </InputGroup>
            <br />

            <InputGroup>
              <InputGroupAddon>Expense Ratio</InputGroupAddon>
              <Input disabled={this.isRootNode()} />
            </InputGroup>
            <br />

            <InputGroup>
              <InputGroupAddon>Cost Basis</InputGroupAddon>
              <Input disabled={this.isRootNode()} />
            </InputGroup>
          </CardBlock>
        </Card>
      </div>
    );
  }
}

NodeDetails.propTypes = {
  selectedNode: PropTypes.object,
  updateDetailsNode: PropTypes.func.isRequired,
};

const mapStateToProps = state => {
  return {
    selectedNode: state.allocationTree.selectedNode,
  };
};

export default connect(mapStateToProps, { updateDetailsNode })(NodeDetails);
