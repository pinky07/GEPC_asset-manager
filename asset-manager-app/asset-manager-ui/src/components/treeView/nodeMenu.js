import { ContextMenu, Item } from 'react-contexify';
import { connect } from 'react-redux';
import { Input } from 'reactstrap';
import PropTypes from 'prop-types';
import React from 'react';

import {
  addAboveNode,
  addBelowNode,
  addSiblingNode,
  deleteBelowNode,
  deleteNode,
  jumpLevel,
} from '../../actions';
import TreeService from '../../services/treeService';

export class NodeMenu extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      jumpLevel: 0,
    };
  }

  componentWillReceiveProps(nextProps) {
    if (!nextProps.selectedNode) {
      return;
    }
    if (nextProps.selectedNode !== this.props.selectedNode) {
      this.setState({
        jumpLevel: nextProps.selectedNode.maxDepth,
      });
    }
  }

  isRootNode = () => TreeService().isRootNode(this.props.selectedNode);

  onChangeInput = event => {
    const max = this.props.selectedNode.maxDepth;
    let jumpLevel = parseInt(event.target.value, 10);

    jumpLevel = jumpLevel > max ? max : jumpLevel;
    jumpLevel = jumpLevel < 1 ? 1 : jumpLevel;
    this.setState({
      jumpLevel,
    });
  };

  onKeyUp = event => {
    if (event.key === 'Enter') {
      this.props.jumpLevel(this.state.jumpLevel);
      event.target.parentNode.click();
    }
  };

  render() {
    return (
      <ContextMenu id="menu_id">
        <Item onClick={this.props.addAboveNode} disabled={this.isRootNode()}>
          Add a Node Above
        </Item>
        <Item onClick={this.props.addSiblingNode} disabled={this.isRootNode()}>
          Add a Sibling Node
        </Item>
        <Item onClick={this.props.addBelowNode}>Add a Node Below</Item>
        <Item disabled>Copy a Node</Item>
        <Item onClick={this.props.deleteNode} disabled={this.isRootNode()}>
          Delete a Node
        </Item>
        <Item onClick={this.props.deleteBelowNode} disabled={this.isRootNode()}>
          Delete the Children Nodes Below
        </Item>
        <Item
          onClick={() => this.props.jumpLevel(this.state.jumpLevel)}
          disabled={!this.state.jumpLevel}>
          <div className="nodeMenu">
            <span>Jump Node Level</span>
            <Input
              type="number"
              step="1"
              min="1"
              max={this.props.selectedNode && this.props.selectedNode.maxDepth}
              value={this.state.jumpLevel}
              onClick={event => event.stopPropagation()}
              onChange={this.onChangeInput}
              onKeyUp={this.onKeyUp}
            />
          </div>
        </Item>
      </ContextMenu>
    );
  }
}

NodeMenu.propTypes = {
  selectedNode: PropTypes.object,
  jumpLevel: PropTypes.func.isRequired,
  addAboveNode: PropTypes.func.isRequired,
  addSiblingNode: PropTypes.func.isRequired,
  addBelowNode: PropTypes.func.isRequired,
  deleteBelowNode: PropTypes.func.isRequired,
  deleteNode: PropTypes.func.isRequired,
};

const mapStateToProps = state => {
  return {
    selectedNode: state.allocationTree.selectedNode,
  };
};

export default connect(mapStateToProps, {
  addAboveNode,
  addSiblingNode,
  addBelowNode,
  deleteBelowNode,
  deleteNode,
  jumpLevel,
})(NodeMenu);
