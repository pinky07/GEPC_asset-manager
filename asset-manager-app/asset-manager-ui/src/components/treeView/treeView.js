import { Col, Row } from 'reactstrap';
import { connect } from 'react-redux';
import { ContextMenuProvider } from 'react-contexify';
import PropTypes from 'prop-types';
import React from 'react';
import SortableTree from 'react-sortable-tree';

import { getAllocationTree, selectNode, updateTree } from '../../actions';
import Constants from '../../services/constants';
import NodeDetails from './nodeDetails';
import NodeMenu from './nodeMenu';
import TreeNodeRenderer from './treeNodeRenderer';

export class TreeView extends React.Component {
  componentWillReceiveProps(nextProps) {
    const { tree, searchResult } = nextProps;
    if (tree !== this.props.tree) {
      if (!this.props.selectedNode) {
        this.props.selectNode({
          node: tree.data[0],
          treeIndex: 0,
          path: [0],
        });
      }
    }

    if (searchResult !== this.props.searchResult) {
      this.props.selectNode(searchResult);
    }
  }

  treeButtons = rowInfo => {
    const buttons = [
      <ContextMenuProvider key="1" id="menu_id" event="onClick">
        <button onClick={() => this.props.selectNode(rowInfo)}>...</button>
      </ContextMenuProvider>,
      <button
        key="2"
        className="buttonInfoNode"
        onClick={() => this.props.selectNode(rowInfo)}
      />,
    ];

    return buttons;
  };

  render() {
    const { tree } = this.props;
    return (
      <div className="allocationTree">
        <Row>
          <Col lg="8" md="8" sm="12" className="treecontainer">
            <SortableTree
              treeData={tree.data}
              onChange={treeData => this.props.updateTree(treeData)}
              canDrop={({ nextPath }) => nextPath.length > 1}
              maxDepth={Constants.MAX_TREE_DEPTH}
              generateNodeProps={rowInfo => ({
                buttons: this.treeButtons(rowInfo),
              })}
              nodeContentRenderer={TreeNodeRenderer}
            />
            <NodeMenu />
          </Col>
          <Col lg="4" md="4" sm="12">
            <NodeDetails />
          </Col>
        </Row>
      </div>
    );
  }
}

TreeView.propTypes = {
  selectedNode: PropTypes.object,
  tree: PropTypes.object,
  getAllocationTree: PropTypes.func.isRequired,
  selectNode: PropTypes.func.isRequired,
  updateTree: PropTypes.func.isRequired,
  searchResult: PropTypes.object,
};

TreeView.defaultProps = {
  tree: {
    data: [],
  },
};

const mapStateToProps = state => {
  return {
    ...state.allocationTree,
  };
};

export default connect(mapStateToProps, {
  getAllocationTree,
  selectNode,
  updateTree,
})(TreeView);
