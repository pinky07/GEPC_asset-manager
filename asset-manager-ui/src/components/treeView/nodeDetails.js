import {
  Card,
  CardBlock,
  CardTitle,
  Input,
  InputGroup,
  InputGroupAddon,
} from 'reactstrap';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import React from 'react';

import TreeService from '../../services/treeService';
import { updateDetailsNode } from '../../actions';
import ColorPicker from './colorPicker';

export class NodeDetails extends React.Component {
  constructor(props) {
    super(props);
    this.node = {
      policy_value: '',
      actual_mv: '',
      accountgroupshortname: '',
      accountgroupname: '',
    };
    this.state = {
      node: this.node,
    };
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.selectedNode) {
      const { node } = nextProps.selectedNode;
      this.setState({
        node,
      });
    } else {
      this.setState({
        node: this.node,
      });
    }
  }

  onChangeInput = (event, nodeProp) => {
    let node = { ...this.state.node };
    node[nodeProp] = event.target.value;
    if (nodeProp === 'accountgroupname') {
      node.title = node.accountgroupname;
    }
    this.setState({
      node,
    });
  };

  onChangePolicy(event) {
    this.setState({
      node: {
        ...this.state.node,
        policy_value: event.target.value,
      },
    });
  }

  onChangeActual(event) {
    this.setState({
      node: {
        ...this.state.node,
        actual_mv: event.target.value,
      },
    });
  }

  isRootNode = () => TreeService().isRootNode(this.props.selectedNode);

  renderPolicyValue = () =>
    this.state.node.policy_value ? `${this.state.node.policy_value}%` : '';

  onBlurTest = () => this.props.updateDetailsNode(this.state.node);

  changeColorAssignment = color => {
    this.setState(
      {
        node: {
          ...this.state.node,
          colorAssignment: color,
        },
      },
      () => this.props.updateDetailsNode(this.state.node)
    );
  };

  render() {
    return (
      <div className="nodeDetails">
        <Card>
          <CardBlock>
            <CardTitle>Node Details</CardTitle>

            <InputGroup>
              <InputGroupAddon>AA Category</InputGroupAddon>
              <Input
                type="select"
                onChange={event => this.onChangeInput(event, 'aa_category')} // TODO Review this!
                disabled={this.isRootNode()}>
                <option>AA Category 1</option>
                <option>AA Category 2</option>
                <option>AA Category 3</option>
              </Input>
            </InputGroup>

            <InputGroup>
              <InputGroupAddon>AA Benchmark</InputGroupAddon>
              <Input
                type="select"
                onChange={event => this.onChangeInput(event, 'aa_benchmark')} // TODO Review this!
                disabled={this.isRootNode()}>
                <option>AA Benchmark 1</option>
                <option>AA Benchmark 2</option>
                <option>AA Benchmark 3</option>
              </Input>
            </InputGroup>

            <InputGroup>
              <InputGroupAddon>Node Name</InputGroupAddon>
              <Input
                value={this.state.node.accountgroupname}
                onChange={event =>
                  this.onChangeInput(event, 'accountgroupname')}
                disabled={this.isRootNode()}
              />
            </InputGroup>

            <InputGroup>
              <InputGroupAddon>Chart Abbreviation</InputGroupAddon>
              <Input
                value="Chart Abbreviation"
                onChange={event =>
                  this.onChangeInput(event, 'chart_abbreviation')} // TODO Review this!
                disabled={this.isRootNode()}
              />
            </InputGroup>

            <InputGroup>
              <InputGroupAddon>Short Code</InputGroupAddon>
              <Input
                type="select"
                onChange={event => this.onChangeInput(event, 'short_code')} // TODO Review this!
                disabled={this.isRootNode()}>
                <option>Short Code 1</option>
                <option>Short Code 2</option>
                <option>Short Code 3</option>
              </Input>
            </InputGroup>

            <InputGroup>
              <InputGroupAddon>CRM InvProd</InputGroupAddon>
              <Input disabled={this.isRootNode()} />
            </InputGroup>

            <InputGroup>
              <InputGroupAddon>Policy Weight</InputGroupAddon>
              <Input
                value={this.renderPolicyValue()}
                onChange={this.onChangePolicy}
                disabled={this.isRootNode()}
              />
            </InputGroup>

            <InputGroup>
              <InputGroupAddon>Actual Market Value</InputGroupAddon>
              <Input
                value={this.state.node.actual_mv || ''}
                onChange={this.onChangeActual}
                disabled={this.isRootNode()}
              />
            </InputGroup>

            <InputGroup>
              <InputGroupAddon>Color Assignment</InputGroupAddon>
              <div
                className={
                  this.isRootNode()
                    ? 'colorPickerWrapper disabled'
                    : 'colorPickerWrapper'
                }>
                <ColorPicker
                  changeColor={this.changeColorAssignment}
                  disabled={this.isRootNode()}
                />
              </div>
            </InputGroup>

            <InputGroup>
              <InputGroupAddon>Liquidity</InputGroupAddon>
              <Input
                type="select"
                onChange={event => this.onChangeInput(event, 'liquidity')} // TODO Review this!
                disabled={this.isRootNode()}>
                <option>Daily</option>
                <option>Monthly</option>
                <option>Yearly</option>
              </Input>
            </InputGroup>

            <InputGroup>
              <InputGroupAddon>Active/Passive</InputGroupAddon>
              <Input
                type="select"
                onChange={event => this.onChangeInput(event, 'active_passive')} // TODO Review this!
                disabled={this.isRootNode()}>
                <option>Active</option>
                <option>Passive</option>
              </Input>
            </InputGroup>

            <InputGroup>
              <InputGroupAddon>Current Expense Ratio</InputGroupAddon>
              <Input disabled={this.isRootNode()} />
            </InputGroup>

            <InputGroup>
              <InputGroupAddon>Cost Basis</InputGroupAddon>
              <Input disabled={this.isRootNode()} />
            </InputGroup>
          </CardBlock>
        </Card>
      </div>
    );
  }
}

NodeDetails.propTypes = {
  selectedNode: PropTypes.object,
  updateDetailsNode: PropTypes.func.isRequired,
};

const mapStateToProps = state => {
  return {
    selectedNode: state.allocationTree.selectedNode,
  };
};

export default connect(mapStateToProps, { updateDetailsNode })(NodeDetails);
