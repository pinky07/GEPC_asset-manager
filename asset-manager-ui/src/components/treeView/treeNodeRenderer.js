import React, { Component } from 'react';
import { isDescendant } from 'react-sortable-tree';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import Constants from '../../services/constants';
import { toggleNodeAtPath } from '../../actions';

class TreeNodeRenderer extends Component {
  toggleNodeOnGrid = event => {
    const { node, path } = this.props;
    this.props.toggleNodeAtPath(
      { ...node, showOnGrid: event.target.checked },
      path
    );
  };

  renderCheckbox = () => {
    const isRootNode = this.props.treeIndex === 0;
    return !isRootNode ? (
      <input
        type="checkbox"
        onChange={this.toggleNodeOnGrid}
        checked={!!this.props.node.showOnGrid}
      />
    ) : null;
  };

  render() {
    const {
      scaffoldBlockPxWidth,
      toggleChildrenVisibility,
      connectDragPreview,
      connectDragSource,
      isDragging,
      canDrop,
      canDrag,
      node,
      title,
      subtitle,
      draggedNode,
      path,
      treeIndex,
      isSearchMatch,
      isSearchFocus,
      buttons,
      className,
      style,
      didDrop,
      selectedNode,
      isOver, // Not needed, but preserved for other renderers
      parentNode, // Needed for dndManager
      toggleNodeAtPath,
      ...otherProps
    } = this.props;

    const isSelected = selectedNode && selectedNode.treeIndex === treeIndex;

    delete otherProps.startDrag;
    delete otherProps.endDrag;

    const nodeTitle = title || node.title;
    const nodeSubtitle = subtitle || node.subtitle;

    let handle;
    if (canDrag) {
      if (typeof node.children === 'function' && node.expanded) {
        // Show a loading symbol on the handle when the children are expanded
        //  and yet still defined by a function (a callback to fetch the children)
        handle = (
          <div className="loadingHandle">
            <div className="loadingCircle">
              <div className="loadingCirclePoint" />
              <div className="loadingCirclePoint" />
              <div className="loadingCirclePoint" />
              <div className="loadingCirclePoint" />
              <div className="loadingCirclePoint" />
              <div className="loadingCirclePoint" />
              <div className="loadingCirclePoint" />
              <div className="loadingCirclePoint" />
              <div className="loadingCirclePoint" />
              <div className="loadingCirclePoint" />
              <div className="loadingCirclePoint" />
              <div className="loadingCirclePoint" />
            </div>
          </div>
        );
      } else {
        // Show the handle used to initiate a drag-and-drop
        handle = connectDragSource(<div className="moveHandle" />, {
          dropEffect: 'copy',
        });
      }
    }

    const isDraggedDescendant = draggedNode && isDescendant(draggedNode, node);
    const isLandingPadActive = !didDrop && isDragging;

    return (
      <div style={{ height: '100%' }} {...otherProps}>
        {toggleChildrenVisibility &&
          node.children &&
          node.children.length > 0 && (
            <div>
              <button
                type="button"
                aria-label={node.expanded ? 'Collapse' : 'Expand'}
                className={node.expanded ? 'collapseButton' : 'expandButton'}
                style={{ left: -0.5 * scaffoldBlockPxWidth }}
                onClick={() =>
                  toggleChildrenVisibility({
                    node,
                    path,
                    treeIndex,
                  })}
              />

              {node.expanded &&
                !isDragging && (
                  <div
                    style={{ width: scaffoldBlockPxWidth }}
                    className="lineChildren"
                  />
                )}
            </div>
          )}

        <div className="rowWrapper">
          {/* Set the row preview to be used during drag and drop */}
          {connectDragPreview(
            <div
              className={
                'row ' +
                (isLandingPadActive ? 'rowLandingPad ' : '') +
                (isLandingPadActive && !canDrop ? 'rowCancelPad ' : '') +
                (isSearchMatch ? 'rowSearchMatch ' : '') +
                (isSearchFocus ? 'rowSearchFocus ' : '') +
                (isSelected ? 'rowSelected ' : '') +
                (className ? ` ${className}` : '')
              }
              style={{
                opacity: isDraggedDescendant ? 0.5 : 1,
                ...style,
              }}>
              {handle}

              <div
                className={
                  'rowContents ' + (!canDrag ? 'rowContentsDragDisabled' : '')
                }>
                <div className="rowLabel">
                  {this.renderCheckbox()}
                  <span
                    className={
                      'rowTitle ' +
                      (node.subtitle ? 'rowTitleWithSubtitle' : '')
                    }
                    style={{
                      color:
                        node.colorAssignment ||
                        Constants.DEFAULT_COLOR_ASSIGNMENT,
                    }}>
                    {typeof nodeTitle === 'function'
                      ? nodeTitle({
                          node,
                          path,
                          treeIndex,
                        })
                      : nodeTitle}
                  </span>

                  {nodeSubtitle && (
                    <span className="rowSubtitle">
                      {typeof nodeSubtitle === 'function'
                        ? nodeSubtitle({
                            node,
                            path,
                            treeIndex,
                          })
                        : nodeSubtitle}
                    </span>
                  )}
                </div>

                <div className="rowToolbar">
                  {buttons.map((btn, index) => (
                    <div
                      key={index} // eslint-disable-line react/no-array-index-key
                      className="toolbarButton">
                      {btn}
                    </div>
                  ))}
                </div>
              </div>
            </div>
          )}
        </div>
      </div>
    );
  }
}

TreeNodeRenderer.defaultProps = {
  isSearchMatch: false,
  isSearchFocus: false,
  canDrag: false,
  toggleChildrenVisibility: null,
  buttons: [],
  className: '',
  style: {},
  parentNode: null,
  draggedNode: null,
  canDrop: false,
  title: null,
  subtitle: null,
};

TreeNodeRenderer.propTypes = {
  node: PropTypes.shape({
    showOnGrid: PropTypes.bool,
  }).isRequired,
  title: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
  subtitle: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
  path: PropTypes.arrayOf(
    PropTypes.oneOfType([PropTypes.string, PropTypes.number])
  ).isRequired,
  treeIndex: PropTypes.number.isRequired,
  isSearchMatch: PropTypes.bool,
  isSearchFocus: PropTypes.bool,
  canDrag: PropTypes.bool,
  scaffoldBlockPxWidth: PropTypes.number.isRequired,
  toggleChildrenVisibility: PropTypes.func,
  buttons: PropTypes.arrayOf(PropTypes.node),
  className: PropTypes.string,
  style: PropTypes.shape({}),
  toggleNodeAtPath: PropTypes.func.isRequired,

  // Drag and drop API functions
  // Drag source
  connectDragPreview: PropTypes.func.isRequired,
  connectDragSource: PropTypes.func.isRequired,
  parentNode: PropTypes.shape({}), // Needed for dndManager
  isDragging: PropTypes.bool.isRequired,
  didDrop: PropTypes.bool.isRequired,
  draggedNode: PropTypes.shape({}),
  // Drop target
  isOver: PropTypes.bool.isRequired,
  canDrop: PropTypes.bool,
  selectedNode: PropTypes.object,
};

const mapStateToProps = state => {
  return {
    selectedNode: state.allocationTree.selectedNode,
  };
};

export default connect(mapStateToProps, { toggleNodeAtPath })(TreeNodeRenderer);
