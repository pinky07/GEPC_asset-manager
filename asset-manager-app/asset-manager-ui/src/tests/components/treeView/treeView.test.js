import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';

import { TreeView } from '../../../components/treeView/treeView';

const props = {
  tree: {
    name: '',
    data: [],
  },
  selectedNode: {},
  getAllocationTree: jest.fn(),
  saveAllocationTree: jest.fn(),
  selectNode: jest.fn(),
  updateTree: jest.fn(),
};

describe('treeView component', () => {
  let wrapper = shallow(<TreeView {...props} />);

  it('renders correctly', () => {
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
