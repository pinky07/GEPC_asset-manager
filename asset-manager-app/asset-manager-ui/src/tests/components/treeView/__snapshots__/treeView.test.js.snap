// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`treeView component renders correctly 1`] = `
<div
  className="allocationTree"
>
  <Row
    tag="div"
  >
    <Col
      className="treecontainer"
      lg="8"
      md="8"
      sm="12"
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <DragDropContext(ReactSortableTree)
        canDrop={[Function]}
        generateNodeProps={[Function]}
        maxDepth={10}
        nodeContentRenderer={[Function]}
        onChange={[Function]}
        searchFocusOffset={0}
        treeData={Array []}
      />
      <Connect(NodeMenu) />
    </Col>
    <Col
      lg="4"
      md="4"
      sm="12"
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <Connect(NodeDetails) />
    </Col>
  </Row>
</div>
`;
